stl/inc/xutility

```diff
@@ -2005,6 +2004,21 @@ _NODISCARD constexpr size_t size(const _Ty (&)[_Size]) noexcept {
     return _Size;
 }
 
+#if _HAS_CXX20
+// FUNCTION TEMPLATE ssize
+template <class _Container>
+_NODISCARD constexpr auto ssize(const _Container& _Cont)
+    -> common_type_t<ptrdiff_t, make_signed_t<decltype(_Cont.size())>> {

```

Comment:
`make_signed<T>` (and therefore `make_signed_t<T>`) has a precondition "*Requires*: `T` shall be a (possibly cv-qualified) integral type or enumeration but not a `bool` type." ([[meta.trans.sign]](http://eel.is/c++draft/meta.trans.sign#sentence-5)) which WG21 seems to have overlooked when specifying this overload of `ssize`. This overload must either "*Mandates*: `decltype(_Cont.size())` [meets the requirements for `make_signed`'s template argument]" or "*Constraints*: `decltype(_Cont.size())` [meets the requirements for `make_signed`'s template argument]" depending on what behavior we want when someone calls `ssize(foo)` with a `foo` that is both convertible to an array type (so the other overload is viable) and has a member `size()` that returns a type unacceptable to `make_signed` (so this overload is viable but may explode).

I'll file the LWG issue for this, and probably include both wording options, but I think we should decide which of "Mandates" or "Constraints" we want to recommend. Thoughts?

Reply:
> is both convertible to an array type (so the other overload is viable)

That can't get past template argument deduction, which demands that the type actually be an array type. (In the case of classes, being derived is acceptable, but you can't derive from an array type.)

"Mandates" would make the current implementation conforming. I believe that there is no reason to have "Constraints", unless we believe that users will provide `ssize()` that will work. Non-integral `size()` return types seem extremely obscure.

Reply:
> That can't get past template argument deduction

Thank you for pointing out that I'm missing the obvious; that certainly simplifies things.

> Non-integral size() return types seem extremely obscure.

Yes, although http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1522r1.pdf indicates that may change in the future. I'll think about reconciling `ssize` with P1522 when I implement `iota_view`.

Reply:
For posterity: This is [LWG 3300](https://wg21.link/lwg3300).